### DISCUSSION TITLE
Bug with For Each enumeration on x64 Custom Classes

### DISCUSSION THREAD

**Discussion by cristianbuse** [Theadd/ds3x#35] [General]
Hi @Theadd ,

I can see you are [using an enumerator inside ArrayList](https://github.com/Theadd/ds3x/blob/5cea8420d797978ab7455e0340e00f7d57ce950e/src/ds3x.accdb.src/modules/ArrayList.cls#L51-L54). I am not sure you are aware of this [bug](https://stackoverflow.com/questions/63848617/bug-with-for-each-enumeration-on-x64-custom-classes) which took me tens of hours to figure out 3 years ago. You must either use one of the known workarounds (e.g. call a method before iterating a custom collection - if you don't forget) or use something like the ```EnumHelper``` class from the answer.

Since I came across this bug I actually only use the ```EnumHelper``` class with greate results because now I iterate multiple things within a class, not just one and moreover I never need the hidden attribute ever again, except the one in ```EnumHelper```. Here is an example project : [KeyedCollection](https://github.com/cristianbuse/VBA-KeyedCollection) - as you can see I can iterate both [items](https://github.com/cristianbuse/VBA-KeyedCollection/blob/9874aa0a980787470cf998ca7155fe1b736ba8f0/src/KeyedCollection.cls#L282-L291) and [keys](https://github.com/cristianbuse/VBA-KeyedCollection/blob/9874aa0a980787470cf998ca7155fe1b736ba8f0/src/KeyedCollection.cls#L343-L352) using a ```For Each v In c.ItemsEnum``` or ```For Each v In c.KeysEnum```. At work I have a class with 5 such iterators.

I'm not saying that you should use the helper class but just be aware of the bug and know the workarounds.

**Comment by Theadd** [comment]
Hi @cristianbuse ,

I've been doing some tests on this, also, last night I noticed that using the custom `ArrayList` enumerator (in For Each) is 2X times slower than in **mscorlib**'s `ArrayList` and 16X times slower than in a `Collection`.

Reproducing your `ShowBug` in my current environment on the custom `ArrayList` implementation, always throws `Error 92: For loop not initialized` once reaching the `Next v` line for the first time (and last, ofc). Here, `v` was `0` regardless of the values in the list.

When enumerating **mscorlib**'s `ArrayList`, this bug does not exist at all, no errors and LongPtr values at 0. Although, using the `ArrayListEx` wrapper from the older version with mscorlib, where `NewEnum()` returns `IUnknown` from `ArrayList.GetEnumerator` <sup><small>[src](https://referencesource.microsoft.com/#mscorlib/system/collections/arraylist.cs,393)</small></sup> instead of an `IEnumVARIANT`, it correctly enumerate the list but not all those LongPtr were at 0, which could crash eventually. So as you stated in your stackoverflow post, as an `IUnknown` instead of an `IEnumVARIANT` won't help it either.

BTW, your [_"**What is happening**"_](https://stackoverflow.com/questions/63848617/bug-with-for-each-enumeration-on-x64-custom-classes/65736562#65736562) answer in that post is impressive.

Your `EnumHelper` solution works, as also do the other workarounds, which, are not suitable for a public class in a shared library. Sadly, the `EnumHelper` is only applicable to methods without the `VB_UserMemId = -4` attribute and to meet the requirements for a "drop-in replacement" It has to implement that attribute. <sup><small><em>(I also have to change `.GetRange` and probably others to fully meet those)</em></small></sup>

```vb
Public Function Items() As EnumHelper
    With New EnumHelper
        Set .EnumVariant = GetNewEnum(Me, pIndex - 1, 0)
        Set Items = .Self
    End With
End Function
```

On second thoughts, since this custom `ArrayList` doesn't have a backing Collection to get it's built-in `_NewEnum`, as far as I can see right now, this should be equivalent to the code above.

```vba
Public Function Items() As Variant()
    Items = ToArray()
End Function
```

Also, since it drops the call to my current IEnumVARIANT laggy implementation, it's 10X times faster. It just has the little extra overheat of having that call to `ToArray()`, only becoming 1.75X times **slower** than in a Collection and 5X times faster than `mscorlib.ArrayList`.

I'll keep thinking about what to do with the `VB_UserMemId = -4` method, I might just print a message in console warning about this and a suggestion to use the `.Items()` instead in `For Each`.

Thank you! =)

**Comment by cristianbuse** [comment]
Hi @Theadd ,

Thanks for the kind words!

**Comment by cristianbuse** [comment]
Hi @Theadd ,

I know you've implemented the custom iterator class in TwinBasic but I still think [this](https://codereview.stackexchange.com/questions/287926/iterate-internal-array-for-a-vba-class) will be an interesting read for you.
